### PR TITLE
Abs and signum added

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**
Although not all features, like derivative, are implemented

### CODE REVIEW

File: AngouriMath/Core/Antlr/AngouriMath.g [hunk 1]
```diff
@@ -136,6 +136,10 @@ atom returns[Entity value]
     | 'limit(' args = function_arguments ')' { Assert("limit", 3, $args.list.Count); $value = MathS.Limit($args.list[0], $args.list[1], $args.list[2]); }
     | 'limitleft(' args = function_arguments ')' { Assert("limitleft", 3, $args.list.Count); $value = MathS.Limit($args.list[0], $args.list[1], $args.list[2], AngouriMath.Core.ApproachFrom.Left); }
     | 'limitright(' args = function_arguments ')' { Assert("limitright", 3, $args.list.Count); $value = MathS.Limit($args.list[0], $args.list[1], $args.list[2], AngouriMath.Core.ApproachFrom.Right); }
+    | 'signum(' args = function_arguments ')' { Assert("signum", 1, $args.list.Count); $value = MathS.Signum($args.list[0]); }
+    | 'sgn(' args = function_arguments ')' { Assert("sgn", 1, $args.list.Count); $value = MathS.Signum($args.list[0]); }
```
Comment (Happypig375) [inline AngouriMath/Core/Antlr/AngouriMath.g:140]: No `Sgn` for `MathS`?

File: AngouriMath/Core/Entity.cs [hunk 2]
```diff
@@ -521,5 +553,24 @@ public Tensor Inverse()
...
+        }
+        public Entity Signum() => new Signumf(this);
+
+        public partial record Absf(Entity Argument) : Function
+        {
+            public override Entity Replace(Func<Entity, Entity> func) => func(Argument.Replace(func).Abs());
+            protected override Entity[] InitDirectChildren() => new[] { Argument };
+        }
+        public Entity Abs() => new Absf(this);
+
```
Comment (Happypig375) [inline AngouriMath/Core/Entity.cs:573]: Why didn't you copy the `New` above? As `ReferenceEquals` is used, if a new object is returned, the cache will be thrown away.

File: AngouriMath/Functions/Algebra/Derivation.cs [hunk 3]
```diff
@@ -169,5 +191,20 @@ public override Entity Derive(Variable variable)
...
+            // or a piecewise definition
+            public override Entity Derive(Variable variable)
+                => MathS.Derivative(this, variable);
+        }
+
+        public partial record Absf
+        {
+            // TODO: derivative of the absolute function?
+            public override Entity Derive(Variable variable)
+                => MathS.Derivative(this, variable);
```
Comment (Happypig375) [inline AngouriMath/Functions/Algebra/Derivation.cs:207]: Derivative of Abs is Signum

File: AngouriMath/Functions/NumbersExtensions.cs [hunk 4]
```diff
@@ -149,6 +155,10 @@ public static EDecimal Sin(this EDecimal x, EContext context)
             return CalculateSinFromCos(x, cos, consts, context);
         }
 
+        public static EDecimal Signum(this EDecimal x, EContext _)
+            => x.Sign;
```
Comment (Happypig375) [inline AngouriMath/Functions/NumbersExtensions.cs:159]: This extension method should be removed, use the property directly

File: AngouriMath/Functions/Output/ToSympy.cs [hunk 5]
```diff
@@ -123,5 +145,17 @@ namespace AngouriMath
...
+                => $@"sympy.sign({Argument.ToSymPy()})";
+        }
+
+        public partial record Absf
+        {
+            internal override string ToSymPy()
+                => $@"sympy.Abs({Argument.ToSymPy()})";
+        }
     }
 }
```
Comment (Happypig375) [inline AngouriMath/Functions/Output/ToSympy.cs:161]: Capital A?

File: Tests/UnitTests/Algebra/SolveTest/SolveOneEquation.cs [hunk 6]
```diff
@@ -35,7 +33,8 @@ static void AssertRootCount(Set roots, int target)
 
         void TestSolver(Entity expr, int rootCount, Integer? toSub = null, bool testNewton = false)
         {
-            var roots = MathS.Settings.AllowNewton.As(false, () => expr.SolveEquation(x));
+            var roots = MathS.Settings.AllowNewton.As(false /* shouldn't it be testNewton instead of false? */,
```
Comment (Happypig375) [inline Tests/UnitTests/Algebra/SolveTest/SolveOneEquation.cs:36]: Newton Solver is tested via SolveNt below

Comment (Happypig375) [reply Tests/UnitTests/Algebra/SolveTest/SolveOneEquation.cs:36]: The idea is that the number of analytical solutions and numerical solutions should be the same

File: AngouriMath/Core/Entity.cs (hunk 2 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Core/Entity.cs:573]: I guess it should be forced via abstract methods. Ill czech it

File: AngouriMath/Functions/Algebra/Derivation.cs (hunk 3 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Functions/Algebra/Derivation.cs:207]: 🤦

File: AngouriMath/Functions/NumbersExtensions.cs (hunk 4 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Functions/NumbersExtensions.cs:159]: It's fine for consistency.

File: AngouriMath/Functions/Output/ToSympy.cs (hunk 5 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Functions/Output/ToSympy.cs:161]: Ye, it's class in sympy

File: AngouriMath/Core/Antlr/AngouriMath.g (hunk 1 above)
Comment (Happypig375) [reply AngouriMath/Core/Antlr/AngouriMath.g:140]: Then it's inconsistent.

Comment (WhiteBlackGoose) [reply AngouriMath/Core/Antlr/AngouriMath.g:140]: Functions in MathS don't have synonyms, there's no need for. Say, IDESs will always give you a hint how you should write it. Also, using MathS. functions implies you want something explicit => strict code. While if you parse from a string, you don't want a strict, explicit code, so there're auto-insertable operators, synonyms

Comment (Happypig375) [reply AngouriMath/Core/Antlr/AngouriMath.g:140]: Well, it's also called the sign function, so why don't you add 'sign(' to here?

Comment (WhiteBlackGoose) [reply AngouriMath/Core/Antlr/AngouriMath.g:140]: We can add it I guess

File: AngouriMath/Functions/Evaluation/Patterns.cs [hunk 7]
```diff
@@ -461,6 +461,16 @@ internal static Entity FactorizeFactorialMultiplications(Entity expr)
             Minusf(Minusf(var any2, var any1), var any1a) when any1 == any1a => any2 - 2 * any1,
             Minusf(Minusf(var any1, var any2), var any1a) when any1 == any1a => -any2,
 
+            Signumf(Signumf(var any1)) => new Signumf(any1),
+            Absf(Absf(var any1)) => new Absf(any1),
+            Absf(Signumf(var _)) => 1,
+            Signumf(Absf(var _)) => 1,
```
Comment (Happypig375) [inline AngouriMath/Functions/Evaluation/Patterns.cs:467]: Can omit `var` for discards

**Pull request merged**